### PR TITLE
Enable context based plan parametrization

### PR DESCRIPTION
**Pull request by kkaarreell**
This change enables plan parametrization using defined context dimensions, as requested in
  https://github.com/teemtee/tmt/issues/1513

### CODE REVIEW

File: tmt/base.py [hunk 1]
```diff
@@ -916,6 +923,7 @@ def _initialize_data_directory(self):
 
     def _fmf_context(self):
         """ Return combined context from plan data and command line """
+        self._plan_context = self.node.get('context', dict())
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/base.py:936]: Your changes to `_expand_node_data` mean `_fmf_context` is now called from `_expand_node_data`, but `_fmf_context` - called on line 817 - tries to read `_plan_context` which is created at line 834.

Would you mind testing your changes with https://github.com/teemtee/tmt/pull/1588? My patch removes the extra `_plan_context`, and the correct member, `context`, should be ready when your changes take effect.

File: tmt/base.py [hunk 2]
```diff
@@ -838,7 +838,24 @@ def __init__(
...
+            # keys and values
+            list_context_string = [
+                (a,
+                 ','.join(b)) if isinstance(
+                    b,
+                    list) else (
+                    a,
+                    str(b)) for (
+                    a,
+                    b) in self._fmf_context().items()]
```
Comment (happz) [inline tmt/base.py:853]: Hmm, could this be refactored for improved readability? It's hard to follow `a` and `b`, and the indentation is also quite unexpected. Something like this, maybe?

```python
list_context_string = [
    (key, ','.join(value) if isinstance(value, list) else str(value))
    for key, value in self._fmf_context().items()
    ]
```

Comment (kkaarreell) [reply tmt/base.py:853]: I can change variable names but the indentation is according to what autopep8 asked me to do. See
https://github.com/teemtee/tmt/actions/runs/3245488693/jobs/5323087268#step:4:126

File: tmt/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kkaarreell) [reply tmt/base.py]: I have cherry-picked your commit.

File: tmt/base.py (hunk 2 above)
Comment (kkaarreell) [reply tmt/base.py]: Changed

File: tmt/base.py [hunk 3]
```diff
@@ -830,15 +837,24 @@ def __init__(
...
     def _expand_node_data(self, data):
         """ Recursively expand variables in node data """
         if isinstance(data, str):
-            return os.path.expandvars(data)
+            # expand env variables
+            data_expanded_vars = os.path.expandvars(data)
+            # convert context values to string and define env variables using context
+            # keys and values
+            list_context_string = [
+                (key, ','.join(value)) if isinstance(value, list) else (key, str(value))
```
Comment (kkaarreell) [inline tmt/base.py:850]: @happz With normalization in place, is this test and else block still necessary?

File: tmt/base.py (hunk 2 above)
Comment (happz) [reply tmt/base.py]: Understood, variable names would help a lot. Can't do much about the autopep8.

---

(mumbling angrily) Autopep8 should be punished as much as possible for this abomination. It helps absolutely no one, the line length is 99 characters, there's no reason to spread a two-liner across so many lines... It's actually hurting people's ability to understand the code :/

File: tmt/base.py (hunk 3 above)
Comment (happz) [reply tmt/base.py:850]: I don't think so, context should be always `Dict[str, List[str]]`, even if there's just a single value in the list.

Comment (kkaarreell) [reply tmt/base.py]: OK. I have tried to remove it.

Comment (happz) [reply tmt/base.py]: One more thing, `modify_environ()` takes `EnvironmentType` which is a fancy name for `Dict[str, str]`, therefore instead of a list you need to prepare a key/value mapping.

Type annotations of `base.py` are under review, no mypy warning so far :/

File: tmt/base.py [hunk 4]
```diff
@@ -846,10 +846,10 @@ def _expand_node_data(self, data):
             data_expanded_vars = os.path.expandvars(data)
             # convert context values to string and define env variables using context
             # keys and values
-            list_context_string = [
-                (key, ','.join(value)) for (key, value) in self._fmf_context().items()
-                ]
-            with tmt.utils.modify_environ(list_context_string):
+            dict_context_vars = {
+                key: ','.join(value) for (key, value) in self._fmf_context().items()
```
Comment (kkaarreell) [inline tmt/base.py:850]: Last thing I am not sure about is if we should join list items as comma separated or space separated.

File: tmt/base.py (hunk 3 above)
Comment (kkaarreell) [reply tmt/base.py]: @happz Updated. Is it correct now?

Comment (happz) [reply tmt/base.py]: Looks correct to me.

**Comment by kkaarreell** [comment]
@psss HI, was this closed by mistake? I do not see it being applied in main.

File: docs/examples.rst [hunk 5]
```diff
@@ -634,20 +634,32 @@ package then could be done in the following way::
...
 
     $ tmt run -e REPO=tmt
 
+For context parametrization the syntax is ``$@dimension`` or
+``$@{dimension}``. The values are set according to the defined
+context specified using ``--context`` command line option and the
+``context`` plan attribute::
+
+    context:
+        repo: tmt
```
Comment (lukaszachy) [inline docs/examples.rst:660]: Would be nice to have example where `$@{repo}` is used.

**Review by lukaszachy — COMMENTED** [review]
One small note about documentation, otherwise LGTM

File: docs/examples.rst (hunk 5 above)
Comment (kkaarreell) [reply docs/examples.rst]: Updated

File: tests/discover/parametrize.sh [hunk 6]
```diff
@@ -27,15 +30,33 @@ rlJournalStart
...
+        rlRun "tmt -c repo=tmt run -r $plan_noctx $steps | tee output"
+        rlAssertGrep 'url: https://github.com/teemtee/tmt' 'output'
+        # Precedence of option over context attribute
+        rlRun "tmt -c repo=fmf run -r $plan_ctx $steps | tee output"
+        rlAssertGrep 'url: https://github.com/teemtee/fmf' 'output'
+    rlPhaseEnd
+
+    rlPhaseStartTest 'Undefined context'
+        rlRun "tmt run -r $plan_noctx $steps | tee output" 2
+        rlAssertGrep 'url: https://github.com/teemtee/${repo}' 'output'
```
Comment (psss) [inline tests/discover/parametrize.sh:57]: So for the undefined context the original content was:

    url: https://github.com/teemtee/$@{repo}

After the expansion it is

    url: https://github.com/teemtee/${repo}

I guess it could be dangerous to modify content if the context dimension name is not matched.

File: tmt/base.py [hunk 7]
```diff
@@ -1001,9 +1001,20 @@ def __init__(
...
+            data_expanded_vars = os.path.expandvars(data)
+            # convert context values to string and define env variables using context
+            # keys and values
+            dict_context_vars = {
+                key: ','.join(value)
+                for (key, value) in self._fmf_context().items()
+                }
+            # convert $@{} to ${} and expand context using temporarily defined
+            # environment variables
+            with tmt.utils.modify_environ(dict_context_vars):
```
Comment (psss) [inline tmt/base.py:1014]: Using the `modify_environment` from within the `_expand_node_data()` method means it can be called many times when more complex dictionaries are process, probably not a big deal, but perhaps it might make sense to handle this simlarly as for `environment`: Update the variables once and do the whole expansion including recursion under it.

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for implementing this. The patch look good in general, however I'm afraid we should not change node data at all when the exact match for context dimension is found.

File: tests/discover/parametrize.sh (hunk 6 above)
Comment (kkaarreell) [reply tests/discover/parametrize.sh]: I believe I have fixed that, although the required code is a bit more complicated.

Comment (kkaarreell) [reply tests/discover/parametrize.sh]: Just to add, previously we could erroneously expand environment variables using context. This should be also fixed now.

File: tmt/base.py [hunk 8]
```diff
@@ -978,7 +978,13 @@ def __init__(
 
         # Store 'environment' and 'environment-file' keys content
         self._environment = self._get_environment_vars(node)
-        # Expand all environment variables in the node
+        # Store context variables in a dictionary se we don't have to
+        # build it over and over again
+        self._dict_context_var = {
```
Comment (happz) [inline tmt/base.py:983]: 1. would it be possible to not store the mapping in self, but rather pass it to method as its second argument? It seem its only purpose is to be consumed by the method, called right after it, never to be used again, and having it as a parameter would remove one shared mutable state, attached to the instance, and the method would be slightly purer.
2. 2. (nitpicking…) WRT the name - a type in the name (“dict”) is very uncommon in python lands, “context” - ther’s so many context, “fmf_ context” would be more fitting.

And by solving 1. we could solve 2. easily, prepare your mapping and pass it tomethod as a “fmf_context” parameter :) No shared state, purer methods, making them easier to test, no loose ends and it’s immediatel clear which context we deal with here :)

Comment (kkaarreell) [reply tmt/base.py]: @happz Updated. I am not sure about type annotations though.

File: tests/discover/parametrize.sh (hunk 6 above)
Comment (kkaarreell) [reply tests/discover/parametrize.sh]: I have added new test scenario covering this.

File: tmt/base.py (hunk 7 above)
Comment (kkaarreell) [reply tmt/base.py]: Updated to optional function argument per @happz 's request.

File: tmt/base.py (hunk 8 above)
Comment (happz) [reply tmt/base.py]: Thank you, it definitely looks better. `Dict[str, str]` is perfectly fine, hard to find a more fitting type annotation. Just two more comments:

* how about making the parameter not optional? AFAICT, seems to be part of each and every call to `_expand_node_data()`, therefore it's not really optional. Giving it a default value does not simplify any call, `fmf_context: Dict[str, str]` would be perfect.
* Not related to the function itself, but one Python trap: never use mutable defaults, like `foo: Dict[str, str] = {}`, as it leads to unexpected behavior. Something to be aware of for the future, more details in https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

Comment (kkaarreell) [reply tmt/base.py]: OK. Updated in the latest commit. Thank you for your feedback.

Comment (happz) [reply tmt/base.py]: Thanks, LGTM!

**Review by psss — APPROVED** [review]
Thanks for improving this, now looks good. Added some minor style adjustments in 81f349f8.
